### ISSUE TITLE
Add field labels when generating toString

### ISSUE THREAD

**Issue by JamesMcIntosh** [felangel/equatable#92]
When you have multiple parameters in the object it becomes hared to distinguish between them.

I would recommend controlling this at a global and class level as done with the `EquatableConfig.stringify` as it can quickly double the amount which ends up in logs.

```
Example(Harry, FF:D5:E4:CF:8E:56, Optional[empty], 22)
```
Would become
```
EquatableConfig.stringifyParameterNames = true;
Example(name: Harry, id: FF:D5:E4:CF:8E:56, phone: Optional[empty], age: 22)
```

**Comment by felangel** [comment]
Hi @JamesMcIntosh 👋 
Thanks for opening an issue!

Unfortunately, I don't think this is possible without requiring code generation or that you explicitly provide the prop names as well. If you have any ideas for how to achieve this I'd love to hear them, thanks 👍

**Comment by JamesMcIntosh** [comment]
Hi @felangel,

To do code generation with the existing code would be difficult as I'm not sure how easy that would be to inspect the contents of the props and compile time and resolve them to a field or method name when generating the toString.

Another way to accomplish this with code generation could be the introduction of an annotation which include / exclude fields to give hints to the code generator but this is a paradigm shift for the library and has flow-on considerations as to whether libraries which generate code should supply them downstream or require consumers to run the codegen.

It seems like the [DataClass](https://pub.dev/packages/dataclass) package is an alternative in this space. 

```
class Person extends Equatable {
  final String name;
  final int age;
  **@ToStringIgnore** final int card;

  Person(this.name, this.age, this.card);

  @override
  List<Object> get props => [name, age, card];
}
```

Your other suggestion of a Map could be done and still maintain backwards compatibility and is not intrusive to the existing code. 

- Add Map of toString properties
- Update mapPropsToString
- Maybe a new `EquatableToString` which extends `Equatable`, makes abstract `toStringProps` and populates props by `toStringProps.values.toList`

I haven't tested the following, it's more a brain-dump.

```
abstract class Equatable {

  Map<String, Object> get toStringProps => null;

  @override
  String toString() {
    switch (stringify) {
      case true:
        return mapPropsToString(runtimeType, props, toStringProps);
      case false:
        return '$runtimeType';
      default:
        return EquatableConfig.stringify == true
            ? mapPropsToString(runtimeType, props, toStringProps)
            : '$runtimeType';
    }
  }

}

String mapPropsToString(Type runtimeType, List<Object> props, Map<String, Object> toStringProps) {
  if (toStringProps != null) {
    return '$runtimeType${toStringProps.map((v) => '${v.key}: ${v.value.toString()}')}';
  } else {
    return '$runtimeType${props?.map((prop) => prop.toString()) ?? '()'}';
  }
}

abstract class EquatableToString extends Equatable {

  @override
  Map<String, Object> toStringProps;

  @override
  List<Object> get props => toStringProps.values.toList;

}

class Person extends Equatable {
  final String name;
  final int age;
  @ToStringIgnore final int card;

  Person(this.name, this.age, this.card);

  @override
  List<Object> get props => toStringProps.values.toList;

  @override
  Map<String, Object> get toStringProps => { 'name': name, 'age': age };
}
```

Kind regards
James

**Comment by SEGVeenstra** [comment]
I've made a fork to try this out, maybe you want to take a look?
https://github.com/SEGVeenstra/equatable

**Comment by JamesMcIntosh** [comment]
@SEGVeenstra Nice work! I like the way you've maintained compatibility. The only thing I'd change is the new lines for each parameter as it would be very chatty on the console.

**Comment by SEGVeenstra** [comment]
> @SEGVeenstra Nice work! I like the way you've maintained compatibility. The only thing I'd change is the new lines for each parameter as it would be very chatty on the console.

Thank you,

I understand about the newlines.
Was more thinking about my own use case 😉.

**Comment by SEGVeenstra** [comment]
@JamesMcIntosh  I've changed it to a single line:
https://github.com/SEGVeenstra/equatable/tree/propsMap

**Comment by felangel** [comment]
@SEGVeenstra @JamesMcIntosh out of curiosity why not just override toString at that point haha?

**Comment by SEGVeenstra** [comment]
> @SEGVeenstra @JamesMcIntosh out of curiosity why not just override toString at that point haha?

🤔

Maybe to get consistency and being able to get the string recursively?

I'll think about it tonight... maybe we are indeed overthinking it.

**Comment by felangel** [comment]
> > @SEGVeenstra @JamesMcIntosh out of curiosity why not just override toString at that point haha?
> 
> 🤔
> 
> Maybe to get consistency and being able to get the string recursively?
> 
> I'll think about it tonight... maybe we are indeed overthinking it.

fwiw you'd still be able to get the string recursively as long as all the props also override `toString` 😄

**Comment by SEGVeenstra** [comment]
> > > @SEGVeenstra @JamesMcIntosh out of curiosity why not just override toString at that point haha?
> > 
> > 🤔
> > 
> > Maybe to get consistency and being able to get the string recursively?
> > 
> > I'll think about it tonight... maybe we are indeed overthinking it.
> 
> fwiw you'd still be able to get the string recursively as long as all the props also override `toString` 😄 

But isn't that like saying, "you don't need Equatable because you can just override == and hashcode"? 😅

The idea is just to make it consise.

**Comment by felangel** [comment]
> But isn't that like saying, "you don't need Equatable because you can just override == and hashcode"? 😅

That's fair but in my opinion the `propsMap` for `toString` is practically the same

```dart
class Person extends Equatable {
  const ({required this.name, required this.age});
  final String name;
  final int age;

  @override
  List<Object> get props => [name, age];

  @override
  Map<String, Object> get propsMap => {'name': name, 'age': age};
}
```

vs

```dart
class Person extends Equatable {
  const ({required this.name, required this.age});
  final String name;
  final int age;

  @override
  List<Object> get props => [name, age];

  @override
  String toString() => 'Person(name: $name, age: $age)`;
}
```

Also, fwiw `runtimeType` will be different in release mode and it won't be the class name in production builds.

**Comment by SEGVeenstra** [comment]
The thing is that in your example you have to override props AND toString.

With propsMap you don't have to include props because it will take them from the propsMap.

So you only implement it once which will prevent you from getting the two out of sync

**Comment by SEGVeenstra** [comment]
> Also, fwiw runtimeType will be different in release mode and it won't be the class name in production builds.

Didn't know that, what will it be in release?

**Comment by SEGVeenstra** [comment]
It would become this:
```dart
class Person extends Equatable {
  const ({required this.name, required this.age});
  final String name;
  final int age;

  @override
  Map<String, Object> get propsMap => {'name': name, 'age': age};
}
```
props would be obsolete, because props can be extracted from the map (like in my fork)

So depending on your needs you could override either props, or propsMap.

**Comment by Azbesciak** [comment]
Hi, any progress on this? The last proposal seems reasonable and natural. The best if dart would support object creation as in js/ts so it could be just `{name, age}` but we have what we have.

**Comment by felangel** [comment]
Closing this as it'll be resolved by #175. You'll be able to use the `@Stringable` macro to achieve this behavior in the future. See also https://github.com/felangel/data_class
